src/dataflow/src/render/sources.rs

```diff
@@ -816,6 +862,148 @@ where
+                        // mentioned above.
+                        .map(|e| e.saturating_sub(1))
+                        .collect::<Vec<_>>(),
+                )
+            })
+        };
+
+        let compaction_handle =
+            BoundedCompactionHandle::new(upper_compaction_bound_fn, compaction_handle);
+
+        let compaction_token: Rc<Box<dyn CompactionHandle>> = Rc::new(Box::new(compaction_handle));
+
+        let previous = self
+            .compaction_handles
+            .insert(source_id.clone(), Rc::downgrade(&compaction_token));
+
+        match previous {
+            Some(compaction_handle) => {
+                if compaction_handle.upgrade().is_some() {
+                    panic!(
+                        "cannot have multiple rendered instances for persisted source {} ({})",
+                        source_id, source_name
+                    );
+                }
+            }
+            None => (), // All good!
+        }
+
+        compaction_token as Rc<dyn Any>
+    }
+
+    /// Enables compaction of sources associated with the identifier.
+    pub fn allow_compaction(&mut self, id: GlobalId, frontier: Antichain<Timestamp>) {
+        if let Some(compaction_handle) = self.compaction_handles.get(&id) {
+            let compaction_handle = match compaction_handle.upgrade() {
+                Some(handle) => handle,
+                None => {
+                    // Clean up the expired weak handle.
+                    // TODO: A better (long term) solution might be the addition of a
+                    // command that tells us to drop a source. Right now, sources are
+                    // being droppped implicitly when all their consumers are being
+                    // dropped.
+                    self.compaction_handles.remove(&id);
+                    return;
+                }
+            };
+
+            compaction_handle.allow_compaction(frontier);
+        }
+    }
+
+    /// Removes the maintained state for source `id`.
+    pub fn del_source(&mut self, id: &GlobalId) -> bool {
+        self.compaction_handles.remove(&id).is_some()
+    }
+}
+
+/// A handle that allows setting a compaction frontier.
+pub trait CompactionHandle {
+    /// Unblocks compaction for updates at or before `since`.
+    fn allow_compaction(&self, since: Antichain<Timestamp>);
+}
+
+/// A handle that allows setting a compaction frontier that is bounded by an upper bound.
+struct BoundedCompactionHandle {
+    /// Returns the upper bound when asked to.
+    upper_bound_fn: Box<dyn Fn() -> Antichain<Timestamp>>,
+
+    /// Underlying compaction handle.
+    compaction_handle: MultiWriteHandle,
+}
+
+impl BoundedCompactionHandle {
+    /// Creates a new [`BoundedCompactionHandle`] that bounds compaction using the given
+    /// `upper_bound_fn` and forwards compaction requests to the given `compaction_handle`
+    pub fn new<F>(upper_bound_fn: F, compaction_handle: MultiWriteHandle) -> Self
```

Comment:
any particular reason this takes a closure instead of simply owning the `ProbeHandle<Timestamp>`?

Reply:
I had planned to write unit tests for `BoundeCompactionHandle`, and I thought I couldn't easily mock that. I could mock if I wanted, because it just wraps a `MutableAntichain`, and I didn't add tests after all. :see_no_evil: 

I will simplify to use `ProbeHandle`.